File: mkdocs/overrides/devbook/accounts/configure-multisig.mjs
import { PrivateKey } from 'symbol-sdk';
import {
	KeyPair,
	SymbolTransactionFactory,
	models,
	NetworkTimestamp,
	SymbolFacade
} from 'symbol-sdk/symbol';

const NODE_URL = process.env.NODE_URL ||
	'https://reference.symboltest.net:3001';
console.log('Using node', NODE_URL);

// Helper function to announce a transaction
async function announceTransaction(payload, label) {
	console.log(`Announcing ${label} to /transactions`);
	const response = await fetch(`${NODE_URL}/transactions`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	console.log('  Response:', await response.text());
}

// Helper function to wait for transaction confirmation
async function waitForConfirmation(transactionHash, label) {
	console.log(`Waiting for ${label} confirmation...`);
	for (let attempt = 0; attempt < 60; attempt++) {
		await new Promise(resolve => setTimeout(resolve, 1000));
		try {
			const response = await fetch(
				`${NODE_URL}/transactionStatus/${transactionHash}`);
			const status = await response.json();
			console.log('  Transaction status:', status.group);
			if (status.group === 'confirmed') {
				console.log(`${label} confirmed in`, attempt, 'seconds');
				return;
			}
			if (status.group === 'failed') {
				throw new Error(`${label} failed: ${status.code}`);
			}
		} catch (e) {
			if (e.message.includes('failed'))
				throw e;
			console.log('  Transaction status: unknown');
		}
	}
	throw new Error(`${label} not confirmed after 60 seconds`);
}

// Returns the cosignatory addresses of the provided multisig account,
// or an empty list if the account is not multisig or has never been used
async function getMultisigCosignatories(address) {
	const multisigPath = `/account/${address}/multisig`;
	console.log(`Getting cosignatories from ${multisigPath}`);
	try {
		const response = await fetch(`${NODE_URL}${multisigPath}`);
		const json = await response.json();
		const cosignatories = json.multisig.cosignatoryAddresses;
		console.log('  Response:', JSON.stringify(cosignatories));
		return cosignatories;
	} catch {
		console.log('  Response: No cosignatories');
		return [];
	}
}

// Returns a transaction that turns a regular account into a multisig
function multisigEnableTransaction(timestamp, feeMult) {
	// Create an embedded multisig account modification transaction
	// that adds two cosignatories
	const embeddedTransaction = facade.transactionFactory
		.createEmbedded({
			type: 'multisig_account_modification_transaction_v1',
			// This is the account that will be turned into a multisig
			signerPublicKey: multisigKeyPair.publicKey,
			// Delta of the number of signatures required for approvals
			minApprovalDelta: 1,
			// Delta of the number of signatures required for removals
			minRemovalDelta: 1,
			addressAdditions: cosignatoryAddresses
		});

	// Build the aggregate transaction
	const embeddedTransactions = [embeddedTransaction];
	const transaction = facade.transactionFactory.create({
		type: 'aggregate_complete_transaction_v3',
		// This is the account that will pay for this transaction
		signerPublicKey: multisigKeyPair.publicKey,
		deadline: timestamp.addHours(2).timestamp,
		transactionsHash: facade.static.hashEmbeddedTransactions(
			embeddedTransactions),
		transactions: embeddedTransactions
	});
	// Reserve space for two cosignatures (each is 104 bytes)
	// and calculate fee for the final transaction size
	transaction.fee = new models.Amount(
		feeMult * (transaction.size + 104 * cosignatoryKeyPairs.length));
	console.log('Enabling the multisig with the aggregate transaction:');
	console.log(JSON.stringify(transaction.toJson(), null, 2));

	// Sign the aggregate transaction with the multisig's signature
	SymbolTransactionFactory.attachSignature(transaction,
		facade.signTransaction(multisigKeyPair, transaction));

	// Append signatures from all cosignatories
	for (const cosignatoryKeyPair of cosignatoryKeyPairs) {
		transaction.cosignatures.push(
			facade.cosignTransaction(cosignatoryKeyPair, transaction));
	}

	return transaction;
}

// Returns a transaction that turns a multisig into a regular account
function multisigDisableTransaction(timestamp, feeMult) {
	// Create two embedded multisig account modification transactions
	// because cosignatories must be removed one by one
	const embeddedTransaction1 = facade.transactionFactory
		.createEmbedded({
			type: 'multisig_account_modification_transaction_v1',
			// This is the multisig account that will be modified
			signerPublicKey: multisigKeyPair.publicKey,
			// Keep required signatures unchanged for this step
			minApprovalDelta: 0,
			minRemovalDelta: 0,
			addressDeletions: [cosignatoryAddresses[1]]
		});
	const embeddedTransaction2 = facade.transactionFactory
		.createEmbedded({
			type: 'multisig_account_modification_transaction_v1',
			// This is the multisig account that will be modified
			signerPublicKey: multisigKeyPair.publicKey,
			// Decrease required signatures after final removal
			minApprovalDelta: -1,
			minRemovalDelta: -1,
			addressDeletions: [cosignatoryAddresses[0]]
		});

	// Build the aggregate transaction
	const embeddedTransactions = [embeddedTransaction1,
		embeddedTransaction2];
	const transaction = facade.transactionFactory.create({
		type: 'aggregate_complete_transaction_v3',
		// This is the account that will pay for this transaction
		signerPublicKey: cosignatoryKeyPairs[0].publicKey,
		deadline: timestamp.addHours(2).timestamp,
		transactionsHash: facade.static.hashEmbeddedTransactions(
			embeddedTransactions),
		transactions: embeddedTransactions
	});
	// Calculate fee for the final transaction size
	// (No need to reserve space for cosignatures, as there are none)
	transaction.fee = new models.Amount(feeMult * transaction.size);
	console.log('Disabling the multisig with the aggregate transaction:');
	console.log(JSON.stringify(transaction.toJson(), null, 2));

	// Sign the aggregate transaction using the first cosigner's signature
	SymbolTransactionFactory.attachSignature(transaction,
		facade.signTransaction(cosignatoryKeyPairs[0], transaction));

	return transaction;
}

const facade = new SymbolFacade('testnet');

const KEY_PREFIX = '0'.repeat(63);

// Setup the keys for the multisig account and its two cosignatories
const MULTISIG_PRIVATE_KEY = process.env.MULTISIG_PRIVATE_KEY || (
	KEY_PREFIX + '1');
const multisigKeyPair = new KeyPair(new PrivateKey(MULTISIG_PRIVATE_KEY));
const multisigAddress = facade.network.publicKeyToAddress(
	multisigKeyPair.publicKey);
console.log(`Multisig address: ${multisigAddress}`,
	`(public key ${multisigKeyPair.publicKey})`);

const cosignatoryKeyPairs = [];
const cosignatoryAddresses = [];
for (let i = 0; i < 2; i++) {
	const COSIGNATORY_PRIVATE_KEY =
		process.env[`COSIGNATORY${i}_PRIVATE_KEY`] || (
			KEY_PREFIX + String(i + 2));
	const kp = new KeyPair(new PrivateKey(COSIGNATORY_PRIVATE_KEY));
	cosignatoryKeyPairs.push(kp);
	const addr = facade.network.publicKeyToAddress(kp.publicKey);
	cosignatoryAddresses.push(addr);
	console.log(`Cosignatory ${i} address: ${addr}`,
		`(public key ${kp.publicKey})`);
}

try {
	// Fetch current network time
	const timePath = '/node/time';
	console.log('Fetching current network time from', timePath);
	const timeResponse = await fetch(`${NODE_URL}${timePath}`);
	const timeJSON = await timeResponse.json();
	const timestamp = new NetworkTimestamp(
		timeJSON.communicationTimestamps.receiveTimestamp);
	console.log('  Network time:', timestamp.timestamp,
		'ms since nemesis');

	// Fetch recommended fees
	const feePath = '/network/fees/transaction';
	console.log('Fetching recommended fees from', feePath);
	const feeResponse = await fetch(`${NODE_URL}${feePath}`);
	const feeJSON = await feeResponse.json();
	const medianMult = feeJSON.medianFeeMultiplier;
	const minimumMult = feeJSON.minFeeMultiplier;
	const feeMult = Math.max(medianMult, minimumMult);
	console.log('  Fee multiplier:', feeMult);

	// Get current state of the multisig account and decide which
	// operation to perform
	const cosignatories = await getMultisigCosignatories(multisigAddress);
	let transaction;
	let signerKeyPair;
	if (cosignatories.length === 0) {
		// Enable the multisig
		transaction = multisigEnableTransaction(timestamp, feeMult);
		// This operation must be signed by the multisig account
		signerKeyPair = multisigKeyPair;
	} else {
		// Disable the multisig
		transaction = multisigDisableTransaction(timestamp, feeMult);
		// This operation must be signed by one of the cosigners
		signerKeyPair = cosignatoryKeyPairs[0];
	}
	const payload = SymbolTransactionFactory.attachSignature(
		transaction,
		facade.signTransaction(signerKeyPair, transaction));

	// Announce and wait for confirmation
	const transactionHash =
		facade.hashTransaction(transaction).toString();
	console.log(
		'Built aggregate transaction with hash:', transactionHash);
	await announceTransaction(payload, 'aggregate transaction');
	await waitForConfirmation(transactionHash, 'aggregate transaction');

} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/accounts/configure_multisig.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.Address;
import org.symbol.sdk.symbol.FeeCalculator;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.NetworkTimestamp;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.*;
import org.symbol.sdk.symbol.models.*;

final class ConfigureMultisig {
	private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

	private static final String NODE_URL = System.getenv()
		.getOrDefault("NODE_URL", "https://reference.symboltest.net:3001");

	private static final SymbolFacade FACADE = new SymbolFacade("testnet");

	private static final String KEY_PREFIX = "0".repeat(63);

	private static final KeyPair MULTISIG_KEY_PAIR;

	private static final Address MULTISIG_ADDRESS;

	private static final List<KeyPair> COSIGNATORY_KEY_PAIRS =
		new ArrayList<>();

	private static final List<Address> COSIGNATORY_ADDRESSES =
		new ArrayList<>();

	static {
		// [>step-1]
		final String multisigPrivateKey = System.getenv().getOrDefault(
			"MULTISIG_PRIVATE_KEY", KEY_PREFIX + "1");
		MULTISIG_KEY_PAIR = new KeyPair(
			new CryptoTypes.PrivateKey(multisigPrivateKey));
		MULTISIG_ADDRESS = FACADE.network.publicKeyToAddress(
			MULTISIG_KEY_PAIR.getPublicKey());
		System.out.println("Multisig address: " + MULTISIG_ADDRESS
			+ " (public key " + MULTISIG_KEY_PAIR.getPublicKey() + ")");

		for (int i = 0; 2 > i; ++i) {
			final String cosignatoryPrivateKey = System.getenv()
				.getOrDefault(
					"COSIGNATORY" + i + "_PRIVATE_KEY",
					KEY_PREFIX + (i + 2));
			final KeyPair keyPair = new KeyPair(
				new CryptoTypes.PrivateKey(cosignatoryPrivateKey));
			COSIGNATORY_KEY_PAIRS.add(keyPair);
			final Address address = FACADE.network.publicKeyToAddress(
				keyPair.getPublicKey());
			COSIGNATORY_ADDRESSES.add(address);
			System.out.println("Cosignatory " + i + " address: " + address
				+ " (public key " + keyPair.getPublicKey() + ")");
		}
		// [<step-1]
	}

	private ConfigureMultisig() {
	}

	private static JsonNode fetchJson(final String path)
		throws IOException, InterruptedException {
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + path)).GET().build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());
		return OBJECT_MAPPER.readTree(response.body());
	}

	private static void announceTransaction(
		final String payload,
		final String label)
		throws IOException, InterruptedException {
		System.out.println("Announcing " + label + " to /transactions");
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + "/transactions"))
			.header("Content-Type", "application/json")
			.PUT(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());
		System.out.println("  Response: " + response.body());
	}

	private static void waitForConfirmation(
		final String transactionHash,
		final String label)
		throws IOException, InterruptedException {
		System.out.println("Waiting for " + label + " confirmation...");
		for (int attempt = 0; 60 > attempt; ++attempt) {
			Thread.sleep(1000);
			try {
				final JsonNode status = fetchJson(
					"/transactionStatus/" + transactionHash);
				final String group = status.get("group").asText();
				System.out.println("  Transaction status: " + group);
				if ("confirmed".equals(group)) {
					System.out.println(
						label + " confirmed in " + attempt + " seconds");
					return;
				}
				if ("failed".equals(group))
					throw new IOException(label + " failed: "
						+ status.get("code").asText());
			} catch (final JsonProcessingException ex) {
				System.out.println("  Transaction status: unknown");
			}
		}
		throw new IOException(label + " not confirmed after 60 seconds");
	}

	// [>step-3]
	// Returns the cosignatory addresses of the provided multisig account,
	// or an empty list if the account is not multisig or has
	// never been used
	private static List<String> getMultisigCosignatories(
		final Address address)
		throws IOException, InterruptedException {
		final String multisigPath = "/account/" + address + "/multisig";
		System.out.println("Getting cosignatories from " + multisigPath);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + multisigPath)).GET().build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());
		if (2 != response.statusCode() / 100) {
			System.out.println("  Response: No cosignatories");
			return List.of();
		}

		final JsonNode cosignatories = OBJECT_MAPPER
			.readTree(response.body())
			.get("multisig")
			.get("cosignatoryAddresses");
		System.out.println("  Response: " + cosignatories);

		final List<String> result = new ArrayList<>();
		cosignatories.forEach(addressNode -> result.add(
			addressNode.asText()));
		return result;
	} // [<step-3]

	// Returns a transaction that turns a regular account into a multisig
	private static
	AggregateCompleteTransactionV3 multisigEnableTransaction(
		final NetworkTimestamp timestamp,
		final long feeMultiplier) {
		// [>step-5]
		// Create an embedded multisig account modification transaction
		// that adds two cosignatories
		final var embeddedDescriptor = new
			MultisigAccountModificationTransactionV1Descriptor(1, 1)
				.addressAdditions(COSIGNATORY_ADDRESSES);
		final EmbeddedTransaction embeddedTransaction =
			FACADE.createEmbeddedTransactionFromTypedDescriptor(
				embeddedDescriptor, MULTISIG_KEY_PAIR.getPublicKey());
		// [<step-5]
		// Build the aggregate transaction [>step-6]
		final List<EmbeddedTransaction> embeddedTransactions = List.of(
			embeddedTransaction);
		final AggregateCompleteTransactionV3 transaction =
			createAggregateTransaction(
				embeddedTransactions,
				MULTISIG_KEY_PAIR,
				timestamp,
				feeMultiplier,
				COSIGNATORY_KEY_PAIRS.size());
		System.out.println(
			"Enabling the multisig with the aggregate transaction:");
		System.out.println(toPrettyJson(transaction.toJson()));
		// [<step-6]
		// [>step-7]
		// Sign the aggregate transaction with the multisig's signature
		SymbolTransactionFactory.attachSignature(
			transaction,
			FACADE.signTransaction(MULTISIG_KEY_PAIR, transaction));

		// Append signatures from all cosignatories
		final List<Cosignature> cosignatures = new ArrayList<>();
		for (final KeyPair cosignatoryKeyPair : COSIGNATORY_KEY_PAIRS)
			cosignatures.add(FACADE.cosignTransaction(
				cosignatoryKeyPair, transaction));
		transaction.setCosignatures(cosignatures);
		// [<step-7]
		return transaction;
	}

	// Returns a transaction that turns a multisig into a regular account
	private static
	AggregateCompleteTransactionV3 multisigDisableTransaction(
		final NetworkTimestamp timestamp,
		final long feeMultiplier) {
		// [>step-8]
		// Create two embedded multisig account modification transactions
		// because cosignatories must be removed one by one
		final var embeddedDescriptor1 = new
			MultisigAccountModificationTransactionV1Descriptor(0, 0)
				.addressDeletions(List.of(COSIGNATORY_ADDRESSES.get(1)));
		final EmbeddedTransaction embeddedTransaction1 =
			FACADE.createEmbeddedTransactionFromTypedDescriptor(
				embeddedDescriptor1, MULTISIG_KEY_PAIR.getPublicKey());
		final var embeddedDescriptor2 = new
			MultisigAccountModificationTransactionV1Descriptor(-1, -1)
				.addressDeletions(List.of(COSIGNATORY_ADDRESSES.get(0)));
		final EmbeddedTransaction embeddedTransaction2 =
			FACADE.createEmbeddedTransactionFromTypedDescriptor(
				embeddedDescriptor2, MULTISIG_KEY_PAIR.getPublicKey());
		// [<step-8]
		// Build the aggregate transaction [>step-9]
		final List<EmbeddedTransaction> embeddedTransactions = List.of(
			embeddedTransaction1, embeddedTransaction2);
		final AggregateCompleteTransactionV3 transaction =
			createAggregateTransaction(
				embeddedTransactions,
				COSIGNATORY_KEY_PAIRS.get(0),
				timestamp,
				feeMultiplier,
				0);
		System.out.println(
			"Disabling the multisig with the aggregate transaction:");
		System.out.println(toPrettyJson(transaction.toJson()));

		SymbolTransactionFactory.attachSignature(
			transaction,
			FACADE.signTransaction(
				COSIGNATORY_KEY_PAIRS.get(0), transaction));
		// [<step-9]
		return transaction;
	}

	private static
	AggregateCompleteTransactionV3 createAggregateTransaction(
		final List<EmbeddedTransaction> embeddedTransactions,
		final KeyPair signerKeyPair,
		final NetworkTimestamp timestamp,
		final long feeMultiplier,
		final int cosignatureCount) {
		final var descriptor = new
			AggregateCompleteTransactionV3Descriptor(
			SymbolFacade.hashEmbeddedTransactions(embeddedTransactions))
			.transactions(embeddedTransactions);
		final Map<String, Object> aggregateMap = descriptor.toMap();
		aggregateMap.put("signerPublicKey", signerKeyPair.getPublicKey());
		aggregateMap.put(
			"deadline", timestamp.addSeconds(7200).timestamp);
		final AggregateCompleteTransactionV3 transaction =
			(AggregateCompleteTransactionV3) FACADE.transactionFactory
				.create(aggregateMap);
		transaction.setFee(new Amount(
			FeeCalculator.calculateTransactionFee(
				transaction, feeMultiplier, cosignatureCount)));
		return transaction;
	}

	private static String toPrettyJson(final Object value) {
		try {
			return OBJECT_MAPPER.writerWithDefaultPrettyPrinter()
				.writeValueAsString(value);
		} catch (final JsonProcessingException ex) {
			return value.toString();
		}
	}

	public static void main(final String[] args) {
		System.out.println("Using node " + NODE_URL);

		try {
			// Fetch current network time [>step-2]
			final String timePath = "/node/time";
			System.out.println("Fetching current network time from "
				+ timePath);
			final JsonNode timeJson = fetchJson(timePath);
			final NetworkTimestamp timestamp = new NetworkTimestamp(
				timeJson.get("communicationTimestamps")
					.get("receiveTimestamp").asLong());
			System.out.println("  Network time: " + timestamp.timestamp
				+ " ms since nemesis");

			// Fetch recommended fees
			final String feePath = "/network/fees/transaction";
			System.out.println("Fetching recommended fees from "
				+ feePath);
			final JsonNode feeJson = fetchJson(feePath);
			final long feeMultiplier = Math.max(
				feeJson.get("medianFeeMultiplier").asLong(),
				feeJson.get("minFeeMultiplier").asLong());
			System.out.println("  Fee multiplier: " + feeMultiplier);
			// [<step-2]
			// [>step-4]
			// Get current state of the multisig account and decide which
			// operation to perform
			final List<String> cosignatories = getMultisigCosignatories(
				MULTISIG_ADDRESS);
			final AggregateCompleteTransactionV3 transaction;
			if (cosignatories.isEmpty())
				transaction = multisigEnableTransaction(
					timestamp, feeMultiplier);
			else
				transaction = multisigDisableTransaction(
					timestamp, feeMultiplier);

			final String payload =
				SymbolTransactionFactory.toJson(transaction);
			// [<step-4]
			// Announce and wait for confirmation [>step-10]
			final String transactionHash =
				FACADE.hashTransaction(transaction).toString();
			System.out.println(
				"Built aggregate transaction with hash: "
					+ transactionHash);
			announceTransaction(payload, "aggregate transaction");
			waitForConfirmation(transactionHash, "aggregate transaction");
			// [<step-10]
		} catch (final IOException | InterruptedException ex) {
			System.out.println(ex.getMessage());
		}
	}
}
